pkg/config/config.go

```diff
@@ -0,0 +1,361 @@
+package config
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+const (
+	configFileName = "config.json"
+	configDirName  = "obsctl"
+)
+
+// getConfigFilePath returns the obsctl config file path or the first string in override argument.
+// Useful for testing.
+func getConfigFilePath(override ...string) string {
+	if len(override) != 0 {
+		return override[0]
+	}
+
+	usrConfigDir, err := os.UserConfigDir()
+	if err != nil {
+		return configFileName
+	}
+
+	return filepath.Join(usrConfigDir, configDirName, configFileName)
+}
+
+func ensureConfigDir() error {
+	if err := os.MkdirAll(path.Dir(getConfigFilePath()), 0700); err != nil {
+		return fmt.Errorf("creating config directory: %w", err)
+	}
+
+	return nil
+}
+
+// Config represents the structure of the configuration file.
+type Config struct {
+	pathOverride []string
+
+	APIs    map[string]APIConfig `json:"apis"`
+	Current struct {
+		API    string `json:"api"`
+		Tenant string `json:"tenant"`
+	} `json:"current"`
+}
+
+// APIConfig represents configuration for an instance of Observatorium.
+type APIConfig struct {
+	URL      string                  `json:"url"`
+	Contexts map[string]TenantConfig `json:"contexts"`
+}
+
+// TenantConfig represents configuration for a tenant.
+type TenantConfig struct {
+	Tenant string      `json:"tenant"`
+	CAFile []byte      `json:"ca"`
+	OIDC   *OIDCConfig `json:"oidc"`
+}
+
+// OIDCConfig represents OIDC auth config for a tenant.
+type OIDCConfig struct {
+	Token *oauth2.Token `json:"token"`
+
+	Audience     string `json:"audience"`
+	ClientID     string `json:"clientID"`
+	ClientSecret string `json:"clientSecret"`
+	IssuerURL    string `json:"issuerURL"`
+}
+
+// Client returns a OAuth2 HTTP client based on the configuration for a tenant.
+func (t *TenantConfig) Client(ctx context.Context, logger log.Logger) (*http.Client, error) {
+	if t.OIDC != nil {
+		provider, err := oidc.NewProvider(ctx, t.OIDC.IssuerURL)
+		if err != nil {
+			return nil, fmt.Errorf("constructing oidc provider: %w", err)
+		}
+
+		ccc := clientcredentials.Config{
+			ClientID:     t.OIDC.ClientID,
+			ClientSecret: t.OIDC.ClientSecret,
+			TokenURL:     provider.Endpoint().TokenURL,
+			Scopes:       []string{"openid", "offline_access"},
```

Comment:
Should we eventually make scopes configurable as well?

Reply:
Yup! @onprem was ideating about adding multiple auth flows to obsctl, as diff users can have diff ones!